File: Python_Books/算法图解/grokking_algorithms_COPY/04_quicksort/lua/05_quicksort.lua
function concat(array1, array2)
    for _, value in pairs(array2) do
        table.insert(array1, value)
    end
    
    return array1
end

function quicksort(array)
    if #array < 2 then
        -- base case, arrays with 0 or 1 element are already "sorted"
        return array
    else
        -- recursive case
        local pivot = array[1]
        local less, greater = {}, {}

        for i = 2, #array do
            if array[i] <= pivot then
                table.insert(less, array[i])
            else
                table.insert(greater, array[i])
            end
        end
        
        return concat(quicksort(less), concat({pivot}, quicksort(greater)))
    end
end

print(table.concat(quicksort({10, 5, 2, 3}), ", "))

File: Python_Books/算法图解/grokking_algorithms_COPY/04_quicksort/javascript/05_quicksort.js
'use strict';

function quicksort(array) {
  if (array.length < 2) {
    // base case, arrays with 0 or 1 element are already "sorted"
    return array;
  } else {
   // recursive case
   let pivot = array[0];
   // sub-array of all the elements less than the pivot
   let less = array.slice(1).filter(function(el) { return el <= pivot; });
   // sub-array of all the elements greater than the pivot
   let greater = array.slice(1).filter(function(el) { return el > pivot; });
   return quicksort(less).concat([pivot], quicksort(greater));
  }
}

console.log(quicksort([10, 5, 2, 3])); // [2, 3, 5, 10]
